server/src/main/java/com/google/musicanalysis/site/OAuthCallbackServlet.java

```diff
@@ -37,40 +37,51 @@ protected void doGet(HttpServletRequest req, HttpServletResponse res)
     }
 
     var sessionOauthState = req.getSession().getAttribute("oauth:state");
+    var sessionOauthService = req.getSession().getAttribute("oauth:service");
 
     if (!state.equals(sessionOauthState)) {
       res.setStatus(401);
       return;
     }
 
+    String clientSecret;
+    String clientId;
+    URI tokenUri;
+
+    if (sessionOauthService.equals("youtube")) {
+      clientSecret = System.getenv().get("YOUTUBE_CLIENT_SECRET");
+      clientId = Constants.YOUTUBE_CLIENT_ID;
+      tokenUri = URI.create("https://oauth2.googleapis.com/token");
+    } else if (sessionOauthService.equals("spotify")) {
+      clientSecret = System.getenv().get("SPOTIFY_CLIENT_SECRET");
+      clientId = Constants.SPOTIFY_CLIENT_ID;
+      tokenUri = URI.create("https://accounts.spotify.com/api/token");
+    } else {
+      res.setStatus(400);
+      return;
+    }
+
     var redirectUri = System.getenv().get("OAUTH_CALLBACK_URI");
-    var spotifyClientSecret = System.getenv().get("SPOTIFY_CLIENT_SECRET");
 
-    var tokenRequestBody = new StringBuilder();
-    tokenRequestBody.append("grant_type=authorization_code");
-    tokenRequestBody.append("&code=");
-    tokenRequestBody.append(URLEncoder.encode(code, "UTF-8"));
-    tokenRequestBody.append("&redirect_uri=");
-    tokenRequestBody.append(URLEncoder.encode(redirectUri, "UTF-8"));
-    tokenRequestBody.append("&client_id=");
-    tokenRequestBody.append(URLEncoder.encode(Constants.SPOTIFY_CLIENT_ID, "UTF-8"));
-    tokenRequestBody.append("&client_secret=");
-    tokenRequestBody.append(URLEncoder.encode(spotifyClientSecret, "UTF-8"));
+    var tokenReqBody = new URLEncodedBuilder()
+      .add("grant_type", "authorization_code")
+      .add("code", code)
+      .add("redirct_uri", redirectUri)
+      .add("client_id", clientId)
+      .add("client_secret", clientSecret);
```

Comment:
Oh is URLEncodedBuilder() a better function than StringBuilder()? (I see below that this is a custom class you have created to make things more convenient.)

Reply:
Not better, but like you said, more convenient 😛